File: Python/String/Levenshtein.py
INF = 10 ** 9
 
def levenshtein(s,t):
    slen = len(s)
    tlen = len(t)
    li = [[INF] * (tlen + 1) for _ in range(slen + 1)]
 
    for i in range(slen + 1):
        li[i][0] = i
    for i in range(tlen + 1):
        li[0][i] = i
 
    for i in range(slen):
        for j in range(tlen):
            cost = 0 if s[i] == t[j] else 1
            li[i + 1][j + 1] = min(li[i][j + 1] + 1,li[i + 1][j] + 1,li[i][j] + cost)
    return li[slen][tlen]

'''verify: https://onlinejudge.u-aizu.ac.jp/problems/DPL_1_E
 
s = input()
t = input()
print(levenshtein(s,t))
'''

File: C++/String/Levenshtein.cpp
int levenshtein(string s, string t) {
	int slen = s.length(), tlen = t.length();
	vector<vector<ll>>v(slen + 1, vector<ll>(tlen + 1, INF));

	for (int i = 0; i <= slen; ++i) { v[i][0] = i; }
	for (int i = 0; i <= tlen; ++i) { v[0][i] = i; }
	for (int i = 0; i < slen; ++i) {
		for (int j = 0; j < tlen; ++j) {
			int cost;
			s[i] == t[j] ? cost = 0 : cost = 1;
			v[i + 1][j + 1] = min({ v[i][j + 1] + 1,v[i + 1][j] + 1,v[i][j] + cost });
		}
	}
	return v[slen][tlen];
}

/*verify: https://onlinejudge.u-aizu.ac.jp/courses/library/7/DPL/1/DPL_1_E
int main() {
	string s, t;
	cin >> s >> t;
	cout << levenshtein(s, t) << endl;
}
*/